File: src/main/java/com/tomatokey/framework/exception/ApiExceptionHandler.java
package com.tomatokey.framework.exception;

import com.tomatokey.framework.utils.ObjectUtils;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.lang.Nullable;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@RestControllerAdvice
public class ApiExceptionHandler extends ResponseEntityExceptionHandler {

    /**
     * 共通のエラー処理
     * このクラス内でハンドリングしたエラーは全てこのメソッドを経由してください
     *
     * @param ex
     * @param body
     * @param headers
     * @param status
     * @param request
     * @return
     */
    @Override
    protected ResponseEntity<Object> handleExceptionInternal(Exception ex, @Nullable Object body, HttpHeaders headers, HttpStatus status, WebRequest request) {
        log.error(ex.getMessage(), ex);

        if (ObjectUtils.isEmpty(body)) {
            final ErrorResponse errorResponse = new ErrorResponse(status, status.name());
            return super.handleExceptionInternal(ex, errorResponse, headers, status, request);
        }

        return super.handleExceptionInternal(ex, body, headers, status, request);
    }

    /**
     * リクエストパラメータのバリデーションでエラーが発生した場合のレスポンスをハンドリングします
     *
     * @param ex
     * @param headers
     * @param status
     * @param request
     * @return
     */
    @Override
    protected ResponseEntity<Object> handleMethodArgumentNotValid(MethodArgumentNotValidException ex, HttpHeaders headers, HttpStatus status, WebRequest request) {
        final List<ValidError> errors = ex.getBindingResult().getFieldErrors().stream()
                .map(ValidError::of)
                .collect(Collectors.toList());
        final ErrorResponse errorResponse = new ErrorResponse(
                status,
                "パラメータが不正です",
                errors
        );
        return handleExceptionInternal(ex, errorResponse, new HttpHeaders(), status, request);
    }

    /**
     * 予期せぬエラーが発生した場合のレスポンスをハンドリングします
     *
     * @param ex
     * @param request
     * @return
     */
    @ExceptionHandler(Exception.class)
    public ResponseEntity<Object> handleUnexpectedException(Exception ex, WebRequest request) {
        return handleExceptionInternal(ex, null, new HttpHeaders(), HttpStatus.INTERNAL_SERVER_ERROR, request);
    }

}

File: src/main/kotlin/com/prototype/framework/exception/ApiExceptionHandler.kt
package com.prototype.framework.exception

import com.prototype.framework.extension.log
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.lang.Nullable
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.context.request.WebRequest
import org.springframework.web.server.ResponseStatusException
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
import java.util.stream.Collectors

@RestControllerAdvice
class ApiExceptionHandler : ResponseEntityExceptionHandler() {
    /**
     * 共通のエラー処理
     * このクラス内でハンドリングしたエラーは全てこのメソッドを経由してください
     *
     * @param ex
     * @param body
     * @param headers
     * @param status
     * @param request
     * @return
     */
    override fun handleExceptionInternal(
        ex: Exception,
        @Nullable body: Any?,
        headers: HttpHeaders,
        status: HttpStatus,
        request: WebRequest
    ): ResponseEntity<Any> {
        log.error(ex.message, ex)
        return body?.let {
            super.handleExceptionInternal(ex, body, headers, status, request)
        } ?: run {
            val errorResponse = ErrorResponse(status, status.name)
            return super.handleExceptionInternal(ex, errorResponse, headers, status, request)
        }
    }

    /**
     * リクエストパラメータのバリデーションでエラーが発生した場合のレスポンスをハンドリングします
     *
     * @param ex
     * @param headers
     * @param status
     * @param request
     * @return
     */
    override fun handleMethodArgumentNotValid(
        ex: MethodArgumentNotValidException,
        headers: HttpHeaders,
        status: HttpStatus,
        request: WebRequest
    ): ResponseEntity<Any> {
        val errors = ex.bindingResult.fieldErrors.stream()
            .map { error: FieldError -> ValidError.of(error) }
            .collect(Collectors.toList())
        val errorResponse = ErrorResponse(
            status,
            "パラメータが不正です",
            errors
        )
        return handleExceptionInternal(ex, errorResponse, HttpHeaders.EMPTY, status, request)
    }

    /**
     * [ResponseStatusException]をthrowした場合のハンドリングをします
     *
     * @param ex
     * @param request
     * @return
     */
    @ExceptionHandler(ResponseStatusException::class)
    fun handleResponseStatusException(ex: ResponseStatusException, request: WebRequest): ResponseEntity<Any> {
        return handleExceptionInternal(ex, null, ex.responseHeaders, ex.status, request)
    }

    /**
     * 予期せぬエラーが発生した場合のレスポンスをハンドリングします
     *
     * @param ex
     * @param request
     * @return
     */
    @ExceptionHandler(Exception::class)
    fun handleUnexpectedException(ex: Exception, request: WebRequest): ResponseEntity<Any> {
        return handleExceptionInternal(ex, null, HttpHeaders.EMPTY, HttpStatus.INTERNAL_SERVER_ERROR, request)
    }
}